python/private/pypi/env_marker_setting.bzl

```
@@ -0,0 +1,265 @@
+    "@platforms//os:ios": "iOS",
+    "@platforms//os:linux": "Linux",
+    "@platforms//os:netbsd": "NetBSD",
+    "@platforms//os:openbsd": "OpenBSD",
+    "@platforms//os:osx": "Darwin",
+    "@platforms//os:windows": "Windows",
+    # The value is empty string if it cannot be determined:
+    # https://docs.python.org/3/library/platform.html#platform.machine
+    "//conditions:default": "",
+}
+
+def env_marker_setting(*, name, expression, **kwargs):
+    """Creates an env_marker setting.
+
+    Args:
+        name: {type}`str` target name
+        expression: {type}`str` the environment marker string to evaluate
+        **kwargs: {type}`dict` additionally common kwargs.
+    """
+    _env_marker_setting(
+        name = name,
+        expression = expression,
+        os_name = select(_os_name_select_map),
+        sys_platform = select(_sys_platform_select_map),
+        platform_machine = select(_platform_machine_select_map),
+        platform_system = select(_platform_system_select_map),
+        **kwargs
+    )
+
+# todo: maybe put all the env into a single target and have a
+# PyPiEnvMarkersInfo provider? Have --pypi_env=//some:target?
+def _env_marker_setting_impl(ctx):
+    # todo: should unify with pep508_env.bzl
+    env = {}
+
+    runtime = ctx.toolchains[TARGET_TOOLCHAIN_TYPE].py3_runtime
+    if runtime.interpreter_version_info:
+        version_info = runtime.interpreter_version_info
+        env["python_version"] = "{major}.{minor}".format(
+            major = version_info.major,
+            minor = version_info.minor,
+        )
+        full_version = _format_full_version(version_info)
+        env["python_full_version"] = full_version
+        env["implementation_version"] = full_version
+    else:
+        env["python_version"] = _get_flag(ctx.attr._python_version_major_minor_flag)
+        full_version = _get_flag(ctx.attr._python_full_version)
+        env["python_full_version"] = full_version
+        env["implementation_version"] = full_version
+
+    # We assume cpython if the toolchain doesn't specify because it's most
+    # likely to be true.
+    env["implementation_name"] = runtime.implementation_name or "cpython"
+    env["os_name"] = ctx.attr.os_name
+    env["sys_platform"] = ctx.attr.sys_platform
+    env["platform_machine"] = ctx.attr.platform_machine
+
+    # The `platform_python_implementation` marker value is supposed to come from
+    # `platform.python_implementation()`, however, PEP 421 introduced
+    # `sys.implementation.name` to replace it. There's now essentially just two
+    # possible values it might have: CPython or PyPy. Rather than add a field to
+    # the toolchain, we just special case the value from
+    # `sys.implementation.name`
+    platform_python_impl = runtime.implementation_name
+    if platform_python_impl == "cpython":
+        platform_python_impl = "CPython"
+    elif platform_python_impl == "pypy":
+        platform_python_impl = "PyPy"
```

```suggestion
    elif platform_python_impl == "pypy":
        platform_python_impl = "PyPy"
    elif platform_python_impl == "rustpython":
        platform_python_impl = "RustPython"
```

Although that is not strictly correct: https://github.com/RustPython/RustPython/blob/cd1c9be0e1a3b59c5d49c9edf593d66538c5e530/Lib/platform.py#L1082

`RustPython` can attempt to be `CPython` if "needed".

I'm fine with leaving it out for now though.